src/charm.py

```diff
@@ -152,11 +153,59 @@ def __init__(self, *args) -> None:
     ###################
 
     def _on_livepatch_relation_changed(self, event) -> None:
-        """Handle peer `relation-changed` event."""
-        self._config_changed(event)
+        """
+        On peer relation changed hook.
+
+        This hook is for the non-leader units to get notified when the state
+        changes. On the leader unit this hook should be ignored to avoid
+        repetitive workload restarts while handling relations. This also means,
+        on the leader unit, whenever the state changes, the update workload
+        method should be called manually.
+        """
+        if self.unit.is_leader():
+            return
+        if not self._state.is_ready():
+            logging.warning("State is not ready")
+            return
+        self._update_workload_configuration(event)
+
+    def _restart_non_leader_units(self):
+        """
+        Cue non-leader units to update their workload configuration.
+
+        This method is meant to be used by action hooks that *do not* update the
+        state. It's because upon updating the state, the non-leader units will
+        get receive the cue via the peer-relation-changed event.
+        """
+        if not self.unit.is_leader() or not self._state.is_ready():
+            return
+        # By incrementing a value in the state, an peer-relation-changed event
+        # will be fired and the non-leader units will respond to it.
+        self._state.restart_cue = 1 + (self._state.restart_cue or 0)
```

Comment:
What about just 
```suggestion
        self._state.restart_cue = !self._state.restart_cue
```

Reply:
I wanted to track the cues during debugging, that's why I made it incremental. And I think it's best to keep this way, because having `restart_cue` as `False` or `True`, can be confusing whenever someone tries to debug the charm. The only issue here is integer overflow (possibly when some crash loop happens), which I'll push a fix for now.

Reply:
Actually, the overflow exception will never happen, because the int type is unbounded in Python 3.